californium-core/src/main/java/org/eclipse/californium/core/CoapClient.java

```diff
@@ -846,7 +849,15 @@ private Request ifNoneMatch(Request request) {
 	private CoapObserveRelation observeAndWait(Request request, CoapHandler handler) {
 		Endpoint outEndpoint = getEffectiveEndpoint(request);
 		CoapObserveRelation relation = new CoapObserveRelation(request, outEndpoint);
-		request.addMessageObserver(new ObserveMessageObserverImpl(handler, relation));
+		// add message observer to get the response.
+		ObserveMessageObserverImpl messageObserver = new ObserveMessageObserverImpl(handler, relation);
+		request.addMessageObserver(messageObserver);
+		// add notification listener to all notification
+		NotificationListener notificationListener = new Adapter(messageObserver, request);
+		outEndpoint.addNotificationListener(notificationListener);
+		// messageobserver should remove this listener when the request is cancelled
+		messageObserver.setNotificationListener(notificationListener);
+		relation.setNotificationListener(notificationListener);
```

Comment:
This is more or less what I propose, I just modify coapClient to keep the previsous behavoir. (But the API is not really adapted to the fail-over/multi instance use case)

The alternative API is:
org.eclipse.californium.core.network.CoapEndpoint.sendRequest(Request)
org.eclipse.californium.core.network.CoapEndpoint.addNotificationListener(NotificationListener)
org.eclipse.californium.core.network.CoapEndpoint.removeNotificationListener(NotificationListener)
org.eclipse.californium.core.network.CoapEndpoint.cancelObservation(byte[])